vaadin-spreadsheet-flow-parent/vaadin-spreadsheet-flow-client/src/main/java/com/vaadin/component/spreadsheet/client/js/SpreadsheetJsApi.java

```diff
@@ -463,15 +469,44 @@ public void setClass(String classNames) {
         }
     }
 
+    private static String getPopupKey(int row, int column) {
+        return row + "_" + column;
+    }
+
     public void setPopups(String raw) {
-        List<PopupButtonState> l = Parser.parseListOfPopupButtons(raw);
-        l.forEach(state -> {
-            String k = state.row + "_" + state.col;
-            if (!popupButtonWidgets.containsKey(k)) {
-                PopupButtonWidget widget;
-                popupButtonWidgets.put(k, widget = new PopupButtonWidget());
-                popupButtonConnectors.put(k, new PopupButtonConnector());
-                popupButtonStates.put(k, state);
+        List<PopupButtonState> popupButtons = Parser
+                .parseListOfPopupButtons(raw);
+
+        var popupButtonsToRemove = new ArrayList<String>();
+
+        for (var state : popupButtonStates.values()) {
+            if (!popupButtons.contains(state)) {
+                var key = getPopupKey(state.row, state.col);
+                popupButtonsToRemove.add(key);
+            }
+        }
+
+        popupButtonsToRemove.forEach(key -> {
+            var popupButton = popupButtonWidgets.get(key);
+            spreadsheetWidget.removePopupButton(popupButton);
+            popupButtonWidgets.remove(key);
+            popupButtonConnectors.remove(key);
+            popupButtonStates.remove(key);
+        });
```

Comment:
This is needed to remove any `PopupButtonWidget` that is no longer on the `popupButtons` list. Not really performant, though. Can be fixed separately together with `Spreadsheet#unRegisterPopupButton`, by calling a new method on the client passing only the buttons to be removed.